docs/rag/ilab-rag-retrieval.md

```diff
@@ -0,0 +1,455 @@
+
+> **ℹ️ Note:** This stack is still under review. The proposed list represents potential candidates based on the current state of discussions.
+
+The following technologies form the foundation of the proposed solution:
+
+* [Haystack](https://haystack.deepset.ai/): Framework for implementing RAG pipelines and applications.
+* [MilvusLite](https://milvus.io/docs/milvus_lite.md): The default vector database for efficient storage and retrieval of embeddings.
+* [Docling](https://github.com/DS4SD/docling): Document processing tool. For more details, refer to William’s blog, [Docling: The missing document processing companion for generative AI](https://www.redhat.com/en/blog/docling-missing-document-processing-companion-generative-ai).
+
+## 3. Design Considerations
+
+* As decided in [PR #165](https://github.com/instructlab/dev-docs/pull/165), functions related to RAG ingestion
+  and retrieval are located in the dedicated folder `src/instructlab/rag`.
+* The solution must minimize changes to existing modules by importing the required functions from the
+  `instructlab.rag` package.
+* The solution must adopt a pluggable design to facilitate seamless integration of additional components:
+  * **Vector stores**: Support all selected implementations (e.g., Milvus).
+  * **Embedding models**: Handle embedding models using the appropriate embedder implementation for the
+    chosen framework (e.g., Haystack).
+  * Consider using factory functions to abstract implementations and enhance code flexibility.
+* Optional dependencies for 3rd party integrations should be defined in `pyproject.toml` and documented for
+  clarity. Users can install optional components with commands like:
+
+  `pip install instructlab[milvus]`
+
+  3rd party dependencies may also be grouped in files such as `requirements/milvus.txt`.
+
+## 4. Future Enhancements
+
+### 4.1 Model Evaluation
+
+**TODO** A separate ADR will be defined.
+
+### 4.2 Advanced RAG retrieval steps
+
+* [Ranking retriever's result][ranking]:
+
+```bash
+ilab model chat --rag --ranking --ranking-top-k=5 --ranking-model=cross-encoder/ms-marco-MiniLM-L-12-v2
+```
+
+* [Query expansion][expansion]:
+
+```bash
+ilab model chat --rag --query-expansion --query-expansion-prompt="$QUERY_EXPANSION_PROMPT" --query-expansion-num-of-queries=5
+```
+
+* Using retrieval strategy:
+
+```bash
+ilab model chat --rag --retrieval-strategy query-expansion --retrieval-strategy-options="prompt=$QUERY_EXPANSION_PROMPT;num_of_queries=5"
+```
+
+* ...
+
+### 4.3 Containerized Indexing Service
+
+Generate a containerized RAG artifact to expose a `/query` endpoint that can serve as an alternative source :
+
+```bash
+ilab rag ingest --build-image --image-name=docker.io/user/my_rag_artifacts:1.0
+```
+
+Then serve it and use it in a chat session:
+
+```bash
+ilab serve --rag-embeddings --image-name=docker.io/user/my_rag_artifacts:1.0 --port 8123
+ilab model chat --rag --retriever-type api --retriever-uri http://localhost:8123
+```
+
```

Comment:
There are many options, but during issues handling, it was found that users either don’t know about or can’t use them correctly, especially when some options need to be used together, like `--retriever-type` and `--retriever-uri`. Only the developers who wrote it might know about this. To make it easier for users to understand, can we emphasize this in the help message or is there a better way to handle it?